File: solutions/303.range-sum-query-immutable/range-sum-query-immutable.java
class NumArray {
    public int[] nums;
    public int[] sums;
    

    public NumArray(int[] nums) {
        this.nums = nums;
        sums = new int[nums.length];
    }
    
    public int sumRange(int i, int j) {
        return getSum(j) - getSum(i) + nums[i];
       
    }
    public int getSum(int i){
        if(i == 0){
            sums[0] = nums[0];
            return sums[0];
        }
        if(sums[i] == 0){
            sums[i] = nums[i] + getSum(i - 1);
        }
        return sums[i];
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */

File: solutions/303.range-sum-query-immutable/range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
    this.sums = new Array(nums.length);
    for(let i = 0 ; i < nums.length; i++){
        if(i === 0) this.sums[i] = nums[i];
        else{
            this.sums[i] = this.sums[i - 1] + nums[i];
        }
    }
    this.sums.unshift(0);
    
};

/** 
 * @param {number} i 
 * @param {number} j
 * @return {number}
 */
NumArray.prototype.sumRange = function(i, j) {
    return this.sums[j + 1] - this.sums[i] ;
    
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = Object.create(NumArray).createNew(nums)
 * var param_1 = obj.sumRange(i,j)
 */